### PR TITLE
Add jax-only local_cost_function machinery

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR adds utilities to define local_cost_functions that will have their gradients and everything defined, which will be used in the future to write an unsupervised learnng driver.

The api is something along the lines of 
```python
@partial(local_cost_function, static_argnums=0, batch_axes=(None, None, 0, 0, 0))
def local_energy(logpsi, pars, vp, mel, v):
    return jax.numpy.sum(mel * jax.numpy.exp(logpsi(pars, vp) - logpsi(pars, v)))

# usage
ma = ...
v = ...
vp, mels = op.get_conn_flattened(v)
loc_vals = local_energy(ma.jax_forward, ma.parameters, vp, mel, v)
_loc_vals,  der_loc_vals = local_costs_and_grads_function(local_energy, ma.jax_forward, ma.parameters, vp, mel, v)

assert loc_vals == _loc_vals
```

I'm not happy about having to explicitly extract the two fields of machine, but i have to find a way to work around that without triggering a lot of recompilation...

@gcarleo, tell me if this seems ok for you, and i'll wrap this up.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
This is now complete.
Notable additions are:
 -  `abstract_operator.get_conn_padded(self, x)` it's a shorthand for calling `get_conn_flattened(x, pad=True)` and then unflattening everything correctly. The code was copy pasted a few times in the lindblad code, and padded inputs are required for all those functions so I thought that it would be useful to put it in the base class.
The code is essentially 
```python
    def get_conn_padded(self, x):
         sections = _np.empty(x.shape[0], dtype=_np.int32)
         x_primes, mels = self.get_conn_flattened(x, sections, pad=True)
          n_primes = sections[0]
         n_visible = x.shape[1]
          x_primes_r = x_primes.reshape(-1, n_primes, n_visible)
         mels_r = mels.reshape(-1, n_primes)
          return x_primes_r, mels_r
```
 -  The der_log_val of the energy that can be used for optimizing LdagL is now defined by using the machinery introduced in this PR, which will also serve as a useful example. Check the definition [here](https://github.com/netket/netket/blob/05153f3b9566283626a9de8ce2d90082db946c07/netket/operator/_der_local_values_jax.py#L13) and how I call the gradient [here](https://github.com/netket/netket/blob/05153f3b9566283626a9de8ce2d90082db946c07/netket/operator/_der_local_values_jax.py#L30)  

 - I renamed `local_cost_function` to `define_ local_cost_function` to be more explicit. `local_cost_function` is now [a function that computes the cost function in batches](https://github.com/netket/netket/blob/05153f3b9566283626a9de8ce2d90082db946c07/netket/operator/_local_cost_functions.py#L60).

Questions still waiting for an answer:
 - This file is in `netket.operators`. Is this the good place? (it's not a perfect fit, but then otherwise where could we put it...?)
 - I'd still like to find a way to accept `machine` itself instead of having to split it into `machine.jax_forward` and `machine._params` but I'm still unsure of the best way to do this...

File: netket/hilbert.py [hunk 1]
```diff
@@ -1,21 +0,0 @@
-from ._C_netket.hilbert import *
-
-
-def Qubit(graph):
```
Comment (gcarleo) [inline netket/hilbert.py:4]: do we still have Qubit?

Comment (PhilipVinc) [reply netket/hilbert.py:4]: [yes](https://github.com/netket/netket/blob/4b4788420c5d1415c0c9a1621ea596b7de635086/netket/hilbert/qubit.py#L5)
I think those two files where never loaded. We probably forgot them when we deleted the old stuff...

**Comment by gcarleo** [comment]
> This file is in netket.operators. Is this the good place? (it's not a perfect fit, but then otherwise where could we put it...?)

I think it's ok... alternative we might create netket.estimators to conceptually separate quantum operators from "local" operators? 

> I'd still like to find a way to make all the functions defined in this PR accept machine itself instead of having to split it into machine.jax_forward and machine._params but I'm still unsure of the best way to do this...

Yes I see that this is potentially an issue with these APIs (that otherwise are great)... Actually I'd need to understand a bit more how you handle all the cases complex->complex real->complex etc when taking only jax_forward here... tomorrow

**Comment by PhilipVinc** [comment]
For the joy of @gcarleo, I reworked this once again:

 - `define_local_cost_function` now takes an additional (optional) keyword argument `outdtype=` (by default `complex`), in line with Jax complex networks. So that, if you *GUARANTEE* that your local cost function is `real` (for example, by taking `jax.numpy.real`) then the calculation of the gradient exploits that information and does less work and returns real/complex gradient (depending on weight type).

- `local_costs_and_grads_function(local_cost_fun, machine, *args)` now takes a full `machine`, which we use to also extract things as the dtype and outdtype.

I also added a few tests to check that most of those different cases actually work...

This is now ready for review
